### ISSUE TITLE
Following the Sample for StructureMap + MediatR Results in Exceptions

### ISSUE THREAD

**Issue by JKSnd** [jbogard/MediatR#194]
We've noticed an issue with the [integration sample of StructureMap and MediatR](https://github.com/jbogard/MediatR/blob/master/samples/MediatR.Examples.StructureMap/Program.cs). Quite often, SM would throw a build exception reporting "Bi-directional dependency relationship detected" when a Mediator was asked to do its thing.

Please see the discussion with Jeremy Miller of Structuremap [Here](https://github.com/structuremap/structuremap/issues/601#issuecomment-328708267) for the in-depth story. [This particular comment](https://github.com/structuremap/structuremap/issues/601#issuecomment-328708267) is where things start getting really productive.

TL;DR: it has to do with the registered methods for the SingleInstanceFactory and MultiInstanceFactory delegates using a build-time IContext for location instead of an IContainer meant for use at run-time. This manifests in highly concurrent scenarios, as something about caching those delegates that capture the IContext isn't thread safe. The solution was to create a wrapper for these methods. [Here's a Gist](https://gist.github.com/JKSnd/7c28077d9cfbcccc50f9cdbd762c775b).

I have two asks:
1) What do you think about updating the sample using the wrapper?
2) What do you think about (for the next major version rollover) having the Mediator collaborate with a proper interface for service location instead of these two raw delegates?

**Comment by lilasquared** [comment]
@JKSnd i got the "Bi-directional dependency relationship detected!" doing something today.  In the details of the message I noticed this line " Instance of `IPipelineBehavior<LogMessageRequest, Unit>`" as I didn't have any pipeline behaviors except for the default ones shipped with MediatR i commented out those lines in the structuremap config and the error went away.  Is the behavior the same for your application?

**Comment by jbogard** [comment]
No, I'm not going to go with a wrapper object or class, if there's something fundamentally wrong with the internals let's fix that.

**Comment by lilasquared** [comment]
@jbogard do you have a good way to reproduce this?  I have a simple project that does it if you would like me to upload it

**Comment by JKSnd** [comment]
@lilasquared could be. Have you tried reintroducing those lines you removed and testing the fix I proposed in [this Gist](https://gist.github.com/JKSnd/7c28077d9cfbcccc50f9cdbd762c775b)?

**Comment by jbogard** [comment]
No I have no way to reproduce it. I suspect it might be some memoization code internally, where I cache which handler type to use?

**Comment by JKSnd** [comment]
@jbogard the issue is that the SingleInstanceFactory and MultiInstanceFactories point to structuremap internals that aren't meant to be cached the way that the Mediator class does.

It would be a breaking change, but I think introducing an interface to provide those factory methods would be the way to do it properly. Then the interface could be implemented to interact with whichever DI library a developer chooses in a way that the DI library works best.

**Comment by JKSnd** [comment]
[This Gist](https://gist.github.com/JKSnd/1bef8e59dd68398fda7cd984d203b368) has some tests that helped me reproduce the issue, myself. Of course, the error jumps around a bit since it is highly dependent on scheduling.

I did try and take advantage of an old MS research project called Chess to add determinism to it, but had little success. :(

**Comment by jbogard** [comment]
What's your lifecycle for IMediator?

**Comment by jbogard** [comment]
Just as an FYI, Mediator doesn't "cache" those delegates, they're injected. The lifecycle of Mediator should be transient at most. Those tests in your gist are parallel all using the same Mediator instance. That should not happen in a real application.

**Comment by JKSnd** [comment]
This is likely our issue. We've been using a singleton lifecycle. Did we overlook some documentation of this recommendation, or is it not explicitly included anywhere yet?

**Comment by jbogard** [comment]
No it's not but the samples should be updated for sure. I had a hard time with the gist, is that MediatR 2.0? 3.0 is all task-based.

**Comment by JKSnd** [comment]
It is 2.0, yes. We'll be updating to 3.0 shortly, however.

I assume the transient lifecycle is recommended because those delegates are injected? Any other nuance we should be aware of?

We had assumed a Singleton would be preferred since that lifecycle is typically the most performant.

**Comment by jbogard** [comment]
Singletons should usually be reserved for the lowest level dependencies,
there are tons of problems of singletons having dependencies. Most
containers have memory leaks if you try.

**Comment by jbogard** [comment]
I made a change to the sample to show exactly how it should be registered (just to avoid confusion).

**Comment by JKSnd** [comment]
Ran my concurrency tests in the style of the sample (explicit transient and again with always unique lifecycles), using the same style of expressions for the factories, and saw the original errors manifest again. Moved the factory delegates into a wrapper again that gets injected with an IContainer and the errors disappeared. The wrapper itself is also transient in these most recent tests.

I think the issue is with using the IContext instead of IContainer.
